0_1_knapsack_problem.py

```diff
@@ -0,0 +1,72 @@
+from typing import List
+
+
+class Solution:
+
+    def knapsack_0_1_problem(self, weights: List[int], values: List[int], capacity: int):
+
+        weights.insert(0, 0)
+        values.insert(0, 0)
+
+        items = len(weights)
+        dp = [[None for cap in range(capacity + 1)] for item in range(items)]
+
+        # No item - Max value 0
+        for cap in range(capacity + 1):
+            dp[0][cap] = 0
+        # No cap - Max value 0
+        for item in range(items):
+            dp[item][0] = 0
+
+        for item in range(1, items):
+            for cap in range(1, capacity + 1):
+                # if the current bag capacity cannot hold the current item
+                # choose the max we could get not choosing the current item
+                # with the same capacity
+                if cap - weights[item] < 0:
+                    dp[item][cap] = dp[item - 1][cap]
+                else:
+                    # for each item I have two options
+                    #                                   choose the item                 don't choose the item
+                    dp[item][cap] = max(values[item] + dp[item - 1][cap - weights[item]], dp[item - 1][cap])
+        return dp[item][cap]
+
+    def knapsack_0_1_problem_brute_force(self, weights: List[int], values: List[int], capacity: int):
+        """
+            // Time Complexity : O(2^n) For each item we have two choices
+            // Space Complexity : O(m/n) 'm' is the capacity, 'n' is the least weight
+        """
```

Comment:
Again m+n. As I mentioned in choin change problem. Also this will be space only if we are considering recursion stack